### ISSUE TITLE
GlassBR self-referential DDs break generated code

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1135]
Some changes were made as part of commit f5ec138336b056dbf64223a8ee2a0e317ec5fa0c that broke the generated code for GlassBR (the changes to DataDefs.hs and IMods.hs are of particular note here). These changes seem to have been made intentionally to remove code-specific symbols from the GlassBR SRS, however they break the code because the DDs now say that, for example, `J` is calculated by the function for calculating `J`. The broken code looks like this:
```
public static double func_J(InputParameters inParams, double q_hat) throws Exception {
        return func_J(q_hat, inParams.AR);
}
```
And the correct code looks like this:
```
public static double func_J(InputParameters inParams, double q_hat) throws Exception {
    return Interpolation.func_interpZ("SDF.txt", inParams.AR, q_hat);
}
```
If I undo the changes from the above commit to DataDefs.hs and IMods.hs, the generated code will be fixed. However, the SRS will break. Here's what the SRS looks like now:
![image](https://user-images.githubusercontent.com/18634568/53449260-6bc7de00-39e7-11e9-9930-0a2e3a5aa9ca.png)
And here is what it will look like if I undo the change to fix the generated code:
![image](https://user-images.githubusercontent.com/18634568/53449364-a16cc700-39e7-11e9-89ae-6c5ae775dca3.png)
Note that the Equation makes no sense and there is a line in the Description that just says "is the interpZ".

I don't think even the current version of the SRS is correct, though it does match manual. My immediate question is whether I should put in the fix for the code at the expense of the SRS, or leave the code broken so that the SRS does not break even more.

**Comment by smiths** [comment]
@bmaclach, in my opinion there is no way to fix this.  The requirements should be abstract, but the code needs to know what interpolation means.  What we need is the "design" that bridges between the two.  I have written a manual version of the Module Interface Specification that connects the SRS to the code:

https://github.com/smiths/caseStudies/blob/master/CaseStudies/glass/docs/Design/MIS/glassbr_mis.pdf

This document has no analogue in Drasil.  We haven't yet decided if this is the approach to use, or something else.  What is certain is that something has to happen between the requirements and the code.  The only reason so much of glassbr works is that the design is so trivial, since many equations are explicit.  The interpolation is more involved, so additional information is needed.

My vote is to keep the code generation working and break the current SRS.  We have plenty of SRS documents, but very little working code.  It would be nice to keep what we have working.  :-)  If you could also create an issue that highlights that the SRS is broken, that would be great.

We are going to need to prioritize our next steps for Drasil.  I'm not sure what the priorities will be, but the move toward incorporating design decisions is something we shouldn't really put off much longer.

**Comment by JacquesCarette** [comment]
I completely agree with @smiths on the need for 'design' to serve as the link between SRS and code, and that it is the root of the problem here.  Interestingly, my instinct was the opposite to his for what to leave broken!  But, now that I read his reasons, I agree.

I am quite glad that this is bringing to the fore the need for the design language. We knew it would come, but it is always better when it comes for a practical reason rather than some abstract 'need'.

**Comment by smiths** [comment]
@JacquesCarette, I agree that it is nice that the need for a design language is coming up organically.  It is also nice that @bmaclach is bringing it up, since he seems the natural candidate to work on this area for his research.  😄

**Comment by bmaclach** [comment]
The decision on how to proceed in this specific situation has been made and was merged with #1138. The bigger problem is documented in #1166, so I'm closing this issue.
